ui: queries screen

vilterp opened the pull request:
This just shows what's in `crdb_internal.node_statement_stats`, which is just what's on the gateway node. No persistence or combining against nodes or anything. May be nice to have this as a simple debug page though, as a base to experiment with as we build out our query perf stuff. cc @dianasaur323 @couchand @piyush-singh @josueeee 

Fixes #26242

![image](https://user-images.githubusercontent.com/7341/40946402-357244be-682b-11e8-8711-536c231ded63.png)

TODO:
- [x] test the endpoint

Commit:
server: add endpoint /_admin/v1/queries

Exposes statement statistics to the UI: what statements have been
executing, basic stats about them, etc.

Release note: None

Commit:
ui: add "Queries" screen, showing recent queries & stats

Just the contents of `crdb_internal.node_statement_statistics` in HTML
table form.

Release note (admin ui change): add queries page showing fingerprints
of incoming statements and basic statistics about them.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24485)
<!-- Reviewable:end -->

BramGruneir commented:
This will be crazy useful!  You should consider adding a SQL formatter and colourer for the SQL column.

Also... need login and permissions for this?

vilterp commented:
Yes, I think this requires login, since there can be values in here AFAIK.

dianasaur323 commented:
wow, nice! Have you read our query analysis specs that outline additional things we'll be including in node statement stats? Eventually, we'll need to surface something that isn't part of crdb_internal for this... cc @knz

vilterp commented on pkg/server/serverpb/admin.proto:
```diff
@@ -548,6 +547,36 @@ message ReplicaMatrixResponse {
+}
+
+message QueriesResponse {
+  message Query {
```
Note: should probably use this existing proto https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/app_stats.proto#L21

…which I didn't notice when first building this.

vilterp commented on pkg/ui/src/util/loader.tsx:
```diff
@@ -0,0 +1,85 @@
+ * The render function is also passsed a `handleRefresh` function, which can
+ * be used as an onClick handler to make a refresh button.
+ */
+class Loader<T> extends React.Component<LoaderProps<T>, LoaderState<T>> {
```
note to reviewers (@couchand, @mrtracy): added this as an experimental alternative to our usual Redux/`CachedDataReducer` song and dance. I think it worked out pretty well, but I'm happy to take this out and use `CachedDataReducer` in this PR for the sake of consistency, until we decide on a more comprehensive data loading refactor (i.e. @mrtracy's sagas+HOC thing)

couchand commented:
Reviewed 4 of 4 files at r1, 2 of 7 files at r5, 1 of 1 files at r6, 3 of 4 files at r7, 2 of 2 files at r8, 3 of 3 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/server/serverpb/admin.proto, line 556 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEQx93JBf7GHHGgy4ZS:-LEQx93JBf7GHHGgy4ZT:b-ys4ks6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2b1859109ae2e22245c52a7fdb2933479a7c01/pkg/server/serverpb/admin.proto#L556)):*
> ```protobuf
>   }
> 
>   enum OrderDir {
> ```

maybe `OrderByDir` to be consistent?

---

*[pkg/server/serverpb/admin.proto, line 566 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-L9HLwGFmPEFsh5jOEFM-r4-566:-LEQx44K0jcRyrRMeMWG:b-c32si0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2b1859109ae2e22245c52a7fdb2933479a7c01/pkg/server/serverpb/admin.proto#L566)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Note: should probably use this existing proto https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/app_stats.proto#L21-L24

…which I didn't notice when first building this.
</blockquote></details>

Sounds like a good idea!

---

*[pkg/ui/src/util/loader.tsx, line 26 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-L9HLwGFmPEFsh5jOEFO-r10-26:-LER0-zD85tdi6tbeE7e:b-5jtecg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/util/loader.tsx#L26)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

note to reviewers (@couchand, @mrtracy): added this as an experimental alternative to our usual Redux/`CachedDataReducer` song and dance. I think it worked out pretty well and I like its simplicity, but I'm happy to take this out and use `CachedDataReducer` in this PR for the sake of consistency, until we decide on a more comprehensive data loading refactor (i.e. @mrtracy's sagas+HOC thing)
</blockquote></details>

experimentation is good but maybe not in master?

---

*[pkg/ui/src/util/loader.tsx, line 37 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER0XqRA_xC4LKRM9wa:-LER0XqRA_xC4LKRM9wb:b-cb2e4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/util/loader.tsx#L37)):*
> ```XML
>   handleRefresh = () => {
>     this.setState({
>       state: "LOADING",
> ```

I don't really like that this is duplicating information stored in the Redux state.

---

*[pkg/ui/src/views/app/components/layoutSidebar/index.tsx, line 116 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEQzYS66K7QacJi7ca9:-LEQzYS66K7QacJi7caA:b-2auu8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/app/components/layoutSidebar/index.tsx#L116)):*
> ```XML
>           <IconLink to="/databases" icon={databasesIcon} title="Databases" activeFor="/database" />
>           <IconLink to="/jobs" icon={jobsIcon} title="Jobs" />
>           <IconLink to={"/queries"} icon={queriesIcon} title={"Queries"} />
> ```

s/{"/"/
s/"}/"/

---

*[pkg/ui/src/views/queries/index.tsx, line 21 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEQzxRo6xRxy25NlnU5:-LEQzxRo6xRxy25NlnU6:bq84y2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/queries/index.tsx#L21)):*
> ```XML
> 
> interface QueriesPageState {
>   sortSetting: SortSetting;
> ```

it would be nice to put this in the url

---

*[pkg/ui/src/views/queries/index.tsx, line 53 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER-lYu2nvHuUXmF1at:-LER-lYu2nvHuUXmF1au:b-spuiqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/queries/index.tsx#L53)):*
> ```XML
>         <QueriesLoader load={getQueries}>
>           {(data, handleRefresh) => {
>             const columns: ColumnDescriptor<Query$Properties>[] = [
> ```

These could be defined way up above to keep this markup clean.

---

*[pkg/ui/src/views/queries/index.tsx, line 78 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER-YEsCRGel3a9ak6s:-LER-YEsCRGel3a9ak6t:bbfgq0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/queries/index.tsx#L78)):*
> ```XML
>             return (
>               <div className="queries-screen">
>                 <button onClick={handleRefresh} style={{ padding: 2, margin: 2 }}>Refresh</button>
> ```

maybe leave this to #24085?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24485)*
<!-- Sent from Reviewable.io -->

Commit:
sql, server: plumb through less-scrubbed queries to endpoint

Adds a new method to the sql package allowing the server package to
access statement statistics before the column and table names have been
scrubbed out of the queries. Previously only the scrubbed queries were
visible to the server package (used in reg cluster reporting).

Release note: None

Commit:
ui: link to queries page from debug index

Release note: None

vilterp commented:
PTAL:
- stripped out unused ordering parameters
- now using existing `StatementStatistics` proto from `app_stats.proto`
- now using `CachedDataReducer`; experimental `Loader` component saved to a branch on my repo
- added a test for the endpoint using a test cluster; hopefully @couchand you can adapt this to testing your fanout code

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1508 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LElAdL42abnRPBQagTU:-LElAdL42abnRPBQagTV:brkef8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/server/admin.go#L1508)):*
> ```Go
> 	ctx context.Context, req *serverpb.QueriesRequest,
> ) (*serverpb.QueriesResponse, error) {
> 	stmtStats := s.server.pgServer.SQLServer.GetStmtStats()
> ```

This no longer goes through `node_statement_stats`. I'm a little torn:

- more efficient not to go through crdb_internal, since that would be unnecessarily roundtripping the stats into SQL datums and back to the same proto-generated `StatementStatistics` struct
- but I like going through SQL in general, so we can use it for authz eventually

Went with the efficient, less-code option in this case `¯\_(ツ)_/¯`

---

*[pkg/server/serverpb/admin.proto, line 566 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-L9HLwGFmPEFsh5jOEFM-r4-566:-LEl0g4307g-CMHkwglD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2b1859109ae2e22245c52a7fdb2933479a7c01/pkg/server/serverpb/admin.proto#L566)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Sounds like a good idea!
</blockquote></details>

Done.

---

*[pkg/ui/src/views/queries/index.tsx, line 21 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEQzxRo6xRxy25NlnU5:-LEl2RxF9VWlBWSqpRqK:b-15d2ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/queries/index.tsx#L21)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

it would be nice to put this in the url
</blockquote></details>

I agree, but would prefer to do this in a followup PR, especially considering that we'd want to keep it consistent with other pages, e.g. jobs, that use the `LocalSettingsReducer`.

---

*[pkg/ui/src/views/queries/index.tsx, line 53 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER-lYu2nvHuUXmF1at:-LEl1lPR1V7mu7xishxW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/queries/index.tsx#L53)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

These could be defined way up above to keep this markup clean.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/queries/index.tsx, line 78 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER-YEsCRGel3a9ak6s:-LEl1mkA9GD0TdE7g4fa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/queries/index.tsx#L78)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

maybe leave this to #24085?
</blockquote></details>

Done.

---

*[pkg/ui/src/util/loader.tsx, line 26 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-L9HLwGFmPEFsh5jOEFO-r10-26:-LEl0m6E-2JFsL0Yqnql:b-kddp6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/util/loader.tsx#L26)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

experimentation is good but maybe not in master?
</blockquote></details>

Done (ripped this out in favor of `CachedDataReducer`), although I like the `Loader` version more! Let's find time to work on this…

---

*[pkg/ui/src/util/loader.tsx, line 37 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER0XqRA_xC4LKRM9wa:-LEl13RN0K7xfjYY8LT_:bx2sp35) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/util/loader.tsx#L37)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I don't really like that this is duplicating information stored in the Redux state.
</blockquote></details>

Redux is not aware of this state. This component bypasses Redux completely.

---

*[pkg/ui/src/views/app/components/layoutSidebar/index.tsx, line 116 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEQzYS66K7QacJi7ca9:-LEl1NE02XVNtk20GGXi:bkjtuoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/views/app/components/layoutSidebar/index.tsx#L116)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

s/{"/"/
s/"}/"/
</blockquote></details>

Good catch. Took this out anyway, since it's a debug page for now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-:-LElACjjCeybAsnvNUc1:b-rk1sw4)*
<!-- Sent from Reviewable.io -->

couchand commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 16 files at r11, 3 of 3 files at r12, 5 of 5 files at r13, 1 of 1 files at r14.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/server/admin.go, line 1508 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LElAdL42abnRPBQagTU:-LEoo57A43KHO9xL0dGR:b1o5qpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/server/admin.go#L1508)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

This no longer goes through `node_statement_stats`. I'm a little torn:

- more efficient not to go through crdb_internal, since that would be unnecessarily roundtripping the stats into SQL datums and back to the same proto-generated `StatementStatistics` struct
- but I like going through SQL in general, so we can use it for authz eventually

Went with the efficient, less-code option in this case `¯\_(ツ)_/¯`
</blockquote></details>

sounds like the pragmatic choice.

as an aside, it would be nice if we could get the authz without incurring the cost.

---

*[pkg/sql/app_stats.go, line 291 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEosbdi1wKItUkrmeQc:-LEosbdi1wKItUkrmeQd:bq3fjw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/sql/app_stats.go#L291)):*
> ```Go
> }
> 
> func (s *sqlStats) getStmtStats(
> ```

maybe add a method `getUnscrubbedStmtStats`?

---

*[pkg/sql/conn_executor.go, line 375 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEosolU0JZRCc9d2pvy:-LEosolU0JZRCc9d2pvz:bw509ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/sql/conn_executor.go#L375)):*
> ```Go
> // GetStmtStats returns the same thing as GetScrubbedStmtStats, except
> // identifiers (e.g. table and column names) aren't scrubbed.
> func (s *Server) GetStmtStats() []roachpb.CollectedStatementStatistics {
> ```

maybe this should be named `GetUnscrubbedStmtStats` or something to make it obvious

---

*[pkg/ui/src/views/queries/index.tsx, line 85 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEorp6E2gNSGFzGHKT9:-LEorp6E2gNSGFzGHKTA:b-k2wvg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/ui/src/views/queries/index.tsx#L85)):*
> ```XML
>   renderQueries() {
>     if (!this.props.queries.data) {
>       // This should really be handled by a loader component.
> ```

touché

---

*[pkg/ui/src/util/loader.tsx, line 26 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-L9HLwGFmPEFsh5jOEFO-r10-26:-LEoq2Q25bpcv-lguwhx:bvuuhgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/util/loader.tsx#L26)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Done (ripped this out in favor of `CachedDataReducer`), although I like the `Loader` version more! Let's find time to work on this…
</blockquote></details>

Let's!

---

*[pkg/ui/src/util/loader.tsx, line 37 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LER0XqRA_xC4LKRM9wa:-LEoq82vAPkwW2_9HRLA:bqvgor0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1366def82e0dd633ae98968a8dfce450fbedd55/pkg/ui/src/util/loader.tsx#L37)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Redux is not aware of this state. This component bypasses Redux completely.
</blockquote></details>

That I'm not sure I like, but that's a discussion for another day.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-:-LEotWG80w5DbkC57jBh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vilterp commented:
TFTR! Added `UnscrubbedStats`.

bors r+

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/app_stats.go, line 291 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEosbdi1wKItUkrmeQc:-LEoxyXN6qeZGJcfG1Na:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/sql/app_stats.go#L291)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

maybe add a method `getUnscrubbedStmtStats`?
</blockquote></details>

Done.

---

*[pkg/sql/conn_executor.go, line 375 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-LEosolU0JZRCc9d2pvy:-LEoxzNL5Rv_wovmdM-5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a89125af31317c7358d6303f5b05de5ed179747/pkg/sql/conn_executor.go#L375)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

maybe this should be named `GetUnscrubbedStmtStats` or something to make it obvious
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24485#-:-LEozDlJ0_3PI3bZKOdC:b-qqr7or)*
<!-- Sent from Reviewable.io -->

vilterp commented:
bors r+
